### PR TITLE
Add read-only access to the children of a `TreeNode`

### PR DESCRIPTION
**Pull request by davep**
Adds a generic `ImmutableSequence` wrapper class (name and location totally up for debate, although the name of the class got the nod in a Slack chat so I suspect that's okay) and then goes on to use that to wrap the internal child list of a `TreeNode`, making it available in a non-destructive way.

See #1398.

### CODE REVIEW

File: src/textual/_collections.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+class ImmutableSequence(Generic[T]):
+    """Class to wrap a sequence of some sort, but not allow modification."""
+
+    def __init__(self, wrap: Iterable[T]) -> None:
+        """Initialise the immutable sequence.
+
+        Args:
+            wrap (Iterable[T]): The iterable value being wrapped.
+        """
+        self._list = list(wrap)
```
Comment (nitzan-shaked) [inline src/textual/_collections.py:18]: wouldn't it be better to just hold a reference to the original sequence? Why copy it? We're not allowing modification anyway.

Comment (nitzan-shaked) [reply src/textual/_collections.py:18]: (and, not to retract from the above: if copying, why not make it tuple() while we're at it, to REALLY disallow modification?)

File: src/textual/_collections.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        return iter(self._list)
+
+    def __len__(self) -> int:
+        return len(self._list)
+
+    def __length_hint__(self) -> int:
+        return len(self)
+
+    def __bool__(self) -> bool:
+        return bool(len(self))
```
[the remark this answers is not in the document]
Comment (davep) [reply src/textual/_collections.py:45]: Yeah, good call, probably makes sense to whittle it down to that (mostly the above is a handover from a slightly earlier version).

File: src/textual/_collections.py (hunk 1 above)
Comment (davep) [reply src/textual/_collections.py:18]: > wouldn't it be better to just hold a reference to the original sequence? Why copy it? We're not allowing modification anyway.

The request for this was, in part, that it be able to take any reasonable iterable while allowing it to be used list-like.

Comment (nitzan-shaked) [reply src/textual/_collections.py]: I'll give it one more shot, then:

1. I was thinking more something like `KeysView` and `ItemsView` from Python's stdlib. They are "aware of perf tradeoffs", so to speak. So here we would have `SequenceView`, and use it where appropriate. There can be also `ImmutableSequence`, but it doesn't mean it has to be used in the particular case of TreeNode's children.
2. How about adding a check in ImmutableSequence's `__init__()`: if `wrap` is an instance of `Sequence` then just store a reference; other convert to tuple.

Apologies for "insisting", I promise this is the last of them, and hope you find the above useful.

Comment (davep) [reply src/textual/_collections.py]: Nah, appreciate it. Of course unlike those views I want this to be a subscriptable object. But yes, you're right, makes more sense to only convert when absolutely needed.

File: src/textual/_collections.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    def __length_hint__(self) -> int:
+        return len(self)
+
+    def __bool__(self) -> bool:
+        return bool(self._wrap)
+
+    def __contains__(self, item: T) -> bool:
+        return item in self._wrap
+
+    def index(self, item: T) -> int:
```
Comment (willmcgugan) [inline src/textual/_collections.py:50]: There are a couple more parameters to `index`. Check a recently merged typing PR for the typing.

**Review by willmcgugan — CHANGES_REQUESTED** [review]
re naming, I think this is essentially a view. Much like the dict views.

Suggest `ImmutableSequenceView` as a name. I don't know if it qualifies as a collection, so I think maybe just name the file `"_immutable_sequence_view.py"` for the moment.

Since the use case is to provide a immutable view on to a collection. It shouldn't copy the data internally. Let's store a reference to the `Sequence` parameter supplied in the constructor.

**Comment by davep** [comment]
> re naming, I think this is essentially a view. Much like the dict views.
> 
> Suggest `ImmutableSequenceView` as a name. I don't know if it qualifies as a collection, so I think maybe just name the file `"_immutable_sequence_view.py"` for the moment.

Makes sense, I'll make those changes.

> Since the use case is to provide a immutable view on to a collection. It shouldn't copy the data internally. Let's store a reference to the Sequence parameter supplied in the constructor.

I'm not sure I quite follow this though. As the code stands doesn't it store a reference if it's an indexable sequence type, and only retains a copy if the incoming sequence is one that can't be indexed, allowing it ti be indexed. Perhaps I've misunderstood what was asked for here?

File: src/textual/_collections.py (hunk 3 above)
Comment (davep) [reply src/textual/_collections.py:50]: Do you mean https://github.com/Textualize/textual/pull/1456?

**Comment by willmcgugan** [comment]
> I'm not sure I quite follow this though. As the code stands doesn't it store a reference if it's an indexable sequence type, and only retains a copy if the incoming sequence is one that can't be indexed, allowing it ti be indexed. Perhaps I've misunderstood what was asked for here?

It's only ever going to be used as a shim on top of a sequence. Can't see any use for creating these for iterators.

File: src/textual/_collections.py (hunk 3 above)
Comment (willmcgugan) [reply src/textual/_collections.py:50]: yup

**Comment by davep** [comment]
> It's only ever going to be used as a shim on top of a sequence. Can't see any use for creating these for iterators.

Ahh, okay, that makes life easier. I only took that into account because you asked that it do so (accept an iterator too) when we spoke about this yesterday. Happy to drop that part.

**Comment by davep** [comment]
Right, that should be those couple or so issues addressed. Thanks.

**Review by willmcgugan — APPROVED** [review]

File: tests/tree/test_tree_node_children.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+from textual.widgets import Tree, TreeNode
+
+def label_of(node: TreeNode[None]):
+    """Get the label of a node as a string"""
+    return str(node.label)
+
+
+def test_tree_node_children() -> None:
+    """A node's children property should act like an immutable list."""
+    CHILDREN=23
```
[the remark this answers is not in the document]
Comment (davep) [reply tests/tree/test_tree_node_children.py:11]: Not by hand on the tests, no. `black` is run by pre-commit of course so if there's a non-black-happy-ness in the above, that would suggest that both pre-commit and CI black checks avoid the unit tests?

Checking:

```sh
$ poetry run black --check .
would reformat devtools/__init__.py
would reformat snapshot_tests/snapshot_apps/horizontal_auto_width.py
would reformat snapshot_tests/snapshot_apps/multiple_css/multiple_css.py
would reformat renderables/test_sparkline.py
would reformat snapshot_tests/snapshot_apps/key_display.py
would reformat renderables/test_text_opacity.py
would reformat test_path.py
would reformat test_focus.py
would reformat test_immutable_sequence_view.py
would reformat test_node_list.py
would reformat renderables/test_underline_bar.py
would reformat test_arrange.py
would reformat test_unmount.py
would reformat test_table.py
would reformat test_dom.py
would reformat test_screens.py
would reformat tree/test_tree_node_label.py
would reformat tree/test_tree_node_children.py
would reformat test_text_backend.py
would reformat test_widget_child_moving.py
would reformat test_reactive.py
would reformat test_widget_mounting.py
would reformat test_widget_removing.py
would reformat test_geometry.py

Oh no! 💥 💔 💥
24 files would be reformatted, 66 files would be left unchanged.
```

Guess none of us are?

[the remark this answers is not in the document]
Comment (willmcgugan) [reply tests/tree/test_tree_node_children.py:11]: I think we may be excluding tests.

Can you check you are also running the latest black specified in dev dependencies.

[the remark this answers is not in the document]
Comment (davep) [reply tests/tree/test_tree_node_children.py:11]:
```sh
$ poetry run black --version
black, 22.10.0 (compiled: yes)
Python (CPython) 3.11.1
```

Which seems to be a later version that specified in poetry.lock?

```
[[package]]
name = "black"
version = "22.8.0"
```

Although weirdly:

```
black = "^22.3.0,<22.10.0"  # macos wheel issue on 22.10
```

from pyproject.toml

**Pull request merged**